### PR TITLE
feat: update JavaScript implementation and add C `ndarray` implementation for `blas/base/isamax`

### PR DESCRIPTION
**Pull request by aman-095**
Progresses #2039.

## Description

> What is the purpose of this pull request?

This pull request:

-   updates the existing Js implementation to support negative strides.
-   adds C `ndarray` API

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/blas/base/isamax/README.md [hunk 1]
```diff
@@ -143,6 +143,129 @@ console.log( idx );
...
+```
+
+#### c_isamax( N, \*X, strideX )
+
+Finds the index of the first element having the maximum absolute value.
+
+```c
+const float x[] = { 4.0f, 2.0f, -3.0f, 5.0f, -1.0f };
+
+CBLAS_INT idx = c_isamax( 5, x, 1 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/README.md]:
```suggestion
int idx = c_isamax( 5, x, 1 );
```

In our examples, we should use an explicit integer dtype.

File: lib/node_modules/@stdlib/blas/base/isamax/README.md [hunk 2]
```diff
@@ -143,6 +143,129 @@ console.log( idx );
...
+```
+
+#### c_isamax_ndarray( N, \*X, strideX, offsetX )
+
+Finds the index of the first element having the maximum absolute value using alternative indexing semantics.
+
+```c
+const float x[] = { 4.0f, 2.0f, -3.0f, 5.0f, -1.0f };
+
+CBLAS_INT idx = c_isamax_ndarray( 5, x, 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/README.md]:
```suggestion
int idx = c_isamax_ndarray( 5, x, 1, 0 );
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 3]
```diff
@@ -18,15 +18,37 @@
...
+* @param N        number of indexed elements
+* @param X        input array
+* @param strideX  X stride length
+* @param offsetX  starting index for X
+* @return         index value
+*/
+CBLAS_INT API_SUFFIX(c_isamax_ndarray)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	X += stdlib_strided_min_view_buffer_index( N, strideX, offsetX ); // adjust array pointer
+	if ( strideX < 0 ) {
+		return N - 1 - API_SUFFIX(cblas_isamax)( N, X, -strideX );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]: After looking at this, I don't think this actually works. It only works if the maximum absolute value occurs only once in an array. If there are two or more equivalent values, for a negative stride, what is returned will be incorrect. In short, flipping the stride has the effect of returning either the first or the last maximum absolute value.

I am not sure there is a workaround when calling into CBLAS.

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c [hunk 4]
```diff
@@ -18,17 +18,51 @@
...
+		sx = -sx;
+		isamaxsub( &N, X, &sx, &idx );
+		return N - idx;
+	}
+	isamaxsub( &N, X, &strideX, &idx );
+	return idx - 1;
+}
+
+/**
+* Finds the index of the first element having the maximum absolute value.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c]:
```suggestion
* Finds the index of the first element having the maximum absolute value using alternative indexing semantics.
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c (hunk 3 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]: I think the only workaround is performing an explicit data copy which reverses an array having a negative stride before calling into CBLAS.

Comment (aman-095) [reply lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]: @kgryte, I have applied the changes you can review it now if anything is incorrect.

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 5]
```diff
@@ -18,15 +18,62 @@
...
+* @param strideX  X stride length
 * @return         index value
 */
-int c_isamax( const int N, const float *X, const int strideX ) {
-	return cblas_isamax( N, X, strideX );
+CBLAS_INT API_SUFFIX(c_isamax)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX ) {
+	CBLAS_INT idx;
+	float *copyX;
+
+    // Allocate memory for an array to store the copied values
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]: @aman-095 You only need to copy if `strideX` is negative.

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c [hunk 6]
```diff
@@ -18,17 +18,56 @@
...
 */
-int c_isamax( const int N, const float *X, const int strideX ) {
-	int idx;
-	isamaxsub( &N, X, &strideX, &idx );
+CBLAS_INT API_SUFFIX(c_isamax)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX ) {
+	CBLAS_INT stride;
+	CBLAS_INT idx;
+	float *copyX;
+
+    copyX = (float *)malloc( N * sizeof(float) ); // Allocate memory for an array to store the copied values
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c]: @aman-095 Mixed tabs and spaces. And same comment. Only copy when the stride is negative.

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 7]
```diff
@@ -18,15 +18,60 @@
...
+* @param strideX  X stride length
 * @return         index value
 */
-int c_isamax( const int N, const float *X, const int strideX ) {
-	return cblas_isamax( N, X, strideX );
+CBLAS_INT API_SUFFIX(c_isamax)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX ) {
+	CBLAS_INT idx;
+	float *copyX;
+
+	if( strideX < 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
	if ( strideX < 0 ) {
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 8]
```diff
@@ -18,15 +18,60 @@
...
+		copyX = (float *)malloc( N * sizeof(float) ); // Allocate memory for an array to store the copied values.
+		if ( copyX == NULL ) {
+			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed for copyA.\n" );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed when copying the input array to a temporary workspace.\n" );
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 9]
```diff
@@ -18,15 +18,60 @@
...
+		}
+		API_SUFFIX(c_scopy)( N, X, strideX, copyX, 1 );
+		idx = API_SUFFIX(cblas_isamax)( N, copyX, 1 );
+		free( copyX ); // Free the allocated memory.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
		free( copyX ); // free allocated memory
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 10]
```diff
@@ -18,15 +18,60 @@
...
+* @param X        input array
+* @param strideX  X stride length
+* @param offsetX  starting index for X
+* @return         index value
+*/
+CBLAS_INT API_SUFFIX(c_isamax_ndarray)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	CBLAS_INT idx;
+	float *copyX;
+
+	if( strideX < 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
	if ( strideX < 0 ) {
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 11]
```diff
@@ -18,15 +18,60 @@
...
+		copyX = (float *)malloc( N * sizeof(float) ); // Allocate memory for an array to store the copied values
+		if ( copyX == NULL ) {
+			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed for copyA.\n" );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed when copying the input array to a temporary workspace.\n" );
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 12]
```diff
@@ -18,15 +18,60 @@
...
+		}
+		API_SUFFIX(c_scopy_ndarray)( N, X, strideX, offsetX, copyX, 1, 0 );
+		idx = API_SUFFIX(cblas_isamax)( N, copyX, 1 );
+		free( copyX ); // Free the allocated memory
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
		free( copyX ); // free allocated memory
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 13]
```diff
@@ -18,15 +18,60 @@
...
+
+	if ( strideX < 0 ) {
+		copyX = (float *)malloc( N * sizeof(float) ); // Allocate memory for an array to store the copied values.
+		if ( copyX == NULL ) {
+			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed when copying the input array to a temporary workspace.\n" );
+		}
+		API_SUFFIX(c_scopy)( N, X, strideX, copyX, 1 );
+		idx = API_SUFFIX(cblas_isamax)( N, copyX, 1 );
+		free( copyX ); // free allocated memory
+		return idx;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
		// Allocate memory for an array to store the copied values:
		copyX = (float *)malloc( N * sizeof(float) );
		if ( copyX == NULL ) {
			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed when copying the input array to a temporary workspace.\n" );
		}
		// Copy the input array to a temporary workspace:
		API_SUFFIX(c_scopy)( N, X, strideX, copyX, 1 );
		
		// Perform operation:
		idx = API_SUFFIX(cblas_isamax)( N, copyX, 1 );
		
		// Free allocated memory:
		free( copyX );
		
		return idx;
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 14]
```diff
@@ -18,15 +18,60 @@
...
+
+	if ( strideX < 0 ) {
+		copyX = (float *)malloc( N * sizeof(float) ); // Allocate memory for an array to store the copied values
+		if ( copyX == NULL ) {
+			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed when copying the input array to a temporary workspace.\n" );
+		}
+		API_SUFFIX(c_scopy_ndarray)( N, X, strideX, offsetX, copyX, 1, 0 );
+		idx = API_SUFFIX(cblas_isamax)( N, copyX, 1 );
+		free( copyX ); // free allocated memory
+		return idx;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
		// Allocate memory for an array to store the copied values:
		copyX = (float *)malloc( N * sizeof(float) );
		if ( copyX == NULL ) {
			API_SUFFIX(c_xerbla)( 1, "isamax", "Memory allocation failed when copying the input array to a temporary workspace.\n" );
		}
		// Copy values to a temporary workspace:
		API_SUFFIX(c_scopy_ndarray)( N, X, strideX, offsetX, copyX, 1, 0 );
		
		// Perform operation:
		idx = API_SUFFIX(cblas_isamax)( N, copyX, 1 );
		
		// Free allocated memory:
		free( copyX );
		
		return idx;
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 15]
```diff
@@ -18,15 +18,74 @@
...
 * @return         index value
 */
-int c_isamax( const int N, const float *X, const int strideX ) {
-	return cblas_isamax( N, X, strideX );
+CBLAS_INT API_SUFFIX(c_isamax)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX ) {
+	CBLAS_INT idx;
+	float *copyX;
+
+	if ( strideX < 0 ) {
+		// Allocate memory for an array to store the copied values:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
		// Allocate memory for a temporary workspace:
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c [hunk 16]
```diff
@@ -18,15 +18,74 @@
...
+* @param strideX  X stride length
+* @param offsetX  starting index for X
+* @return         index value
+*/
+CBLAS_INT API_SUFFIX(c_isamax_ndarray)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	CBLAS_INT idx;
+	float *copyX;
+
+	if ( strideX < 0 ) {
+		// Allocate memory for an array to store the copied values:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_cblas.c]:
```suggestion
		// Allocate memory for a temporary workspace:
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c [hunk 17]
```diff
@@ -18,17 +18,66 @@
...
 * @return         index value
 */
-int c_isamax( const int N, const float *X, const int strideX ) {
-	int idx;
+CBLAS_INT API_SUFFIX(c_isamax)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX ) {
+	CBLAS_INT stride;
+	CBLAS_INT idx;
+	float *copyX;
+
+	if( strideX < 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c]: @aman-095 Would you mind copying over the changes from `isamax_cblas.c` to this file? Thanks!

**Review by kgryte — COMMENTED** [review]
This is looking good. Just need to clean up the `isamax_f.c` file.

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c [hunk 18]
```diff
@@ -18,17 +18,80 @@
...
 * @return         index value
 */
-int c_isamax( const int N, const float *X, const int strideX ) {
-	int idx;
+CBLAS_INT API_SUFFIX(c_isamax)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX ) {
+	CBLAS_INT idx;
+	CBLAS_INT sx;
+	float *copyX;
+
+	if( strideX < 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c]:
```suggestion
	if ( strideX < 0 ) {
```

File: lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c [hunk 19]
```diff
@@ -18,17 +18,80 @@
...
+* @param strideX  X stride length
+* @param offsetX  starting index for X
+* @return         index value
+*/
+CBLAS_INT API_SUFFIX(c_isamax_ndarray)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	CBLAS_INT idx;
+	CBLAS_INT sx;
+	float *copyX;
+
+	if( strideX < 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/isamax/src/isamax_f.c]:
```suggestion
	if ( strideX < 0 ) {
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @aman-095!

**Pull request merged**